spec.md

```diff
@@ -289,6 +290,34 @@ encapsulated within the `data` attribute.
 * Constraints:
   * OPTIONAL
 
+# Minimum Supported Event Size
+
+In order to increase interoperability, all CloudEvent consumers SHOULD accept
+events that follow these rules:
+
+* The number of attributes does not exceed 100. Each index of a `Map` is
+  considered an attribute.
+* All attribute names are 20 characters or less long.
+* All indexes of `Maps` are 20 characters or less long.
+* All `Binary` attributes do not exceed a size of 1KB.
+* All `String` attributes (including String expressions like `URI-reference` and
+  `Timestamp`) do not exceed a size of 1KB.
```

Comment:
Is 1KB sufficient for a URI-reference? I think 2KB [is accepted by many browsers and other software](https://stackoverflow.com/questions/417142/what-is-the-maximum-length-of-a-url-in-different-browsers).

In any case, is there a particular benefit to capping the size of (presumably context) attributes on a per-attribute basis? 100 * (20 + 20 + 1KB + encoding overhead) >= 100KB, which is larger than the total allowed size.

Additionally, I can imagine some attribute uses (like signatures) which could quite possibly consume more than 1KB -- is there some other system limit which is being worked around here?

Reply:
The rationale is that a middleware needs to interpret some context attribute values. E.g. the `source` or `type` could be used for routing. A middleware may allow to perform a Regex on the `source`.

A DoS attacker will try to maximize the computational overhead per request. To do that, the attacker could send the maximum allowed length for a `source` and have a Regex that forces the middleware to do an expensive calculation.
Even if we ignore DoS attacks, if I want to provide a middleware aaS, and I want to be price-competitive, I'd like the computational work per event to be somewhat similar across tenants. E.g. I'd like to charge $0.00001 per request.
However, if one tenant makes me parse a regex on a 60KB source, this is much more expensive than a tenant who has a regex on a small source.
So either I have a very complicated pricing model, I loose money on some tenants, or I have to overcharge everyone for a few tenants doing crazy things.

That is my experience from running a BaaS. I'll most likely not be running a CloudEvents middleware aaS, so I'm just trying to be nice to the guys who'll end up doing that 😉 But - if no one thinks this is necessary - I can also remove it.

---

Yes, 2KB is also fine. I'll change that.

Reply:
https://github.com/google/re2 should [bound the cost of the regular expression matching to be linear in the cost of the string](https://github.com/google/re2/wiki/WhyRE2).

I'm also comfortable with a 2KB limit; looking at a fairly large JWT (not that we would necessarily pass a JWT as a context attribute, but future encryption might look somewhat similar), the size is 1500 bytes: https://stackoverflow.com/questions/39767231/jwt-access-token-public-information-and-size

Reply:
> https://github.com/google/re2 should bound the cost of the regular expression matching to be linear in the cost of the string.

Which only validates point 😉 
If the average event producer sends a source of 50 bytes, and the DoS event producer sends a source of 50 kilobytes, assuming linear cost it is 1000 times more expensive to match the latter event.

> I'm also comfortable with a 2KB limit

👍